miner/worker.go

```diff
@@ -403,177 +243,91 @@ func (w *worker) newWorkLoop(recommit time.Duration) {
-				w.mu.RUnlock()
-
-				txs := make(map[common.Address]types.Transactions)
-				for _, tx := range ev.Txs {
-					acc, _ := types.Sender(w.current.signer, tx)
-					txs[acc] = append(txs[acc], tx)
-				}
-
-				txset := types.NewTransactionsByPriceAndNonce(w.current.signer, txs, w.createTxCmp())
-				tcount := w.current.tcount
-				w.commitTransactions(txset, txFeeRecipient, nil)
-				// Only update the snapshot if any new transactons were added
-				// to the pending block
-				if tcount != w.current.tcount {
-					w.updateSnapshot()
-				}
-			}
-			atomic.AddInt32(&w.newTxs, int32(len(ev.Txs)))
+			ctx, cancel = context.WithCancel(context.Background())
+			wg.Add(1)
+			go func() {
+				w.constructAndSubmitNewBlock(ctx)
+				wg.Done()
+			}()
 
 		// System stopped
 		case <-w.exitCh:
-			return
-		case <-w.txsSub.Err():
+			if cancel != nil {
+				cancel()
+			}
 			return
 		case <-w.chainHeadSub.Err():
-			return
-		case <-w.chainSideSub.Err():
+			if cancel != nil {
+				cancel()
+			}
 			return
 		}
 	}
 }
 
-// taskLoop is a standalone goroutine to fetch sealing task from the generator and
-// push them to consensus engine.
-func (w *worker) taskLoop() {
-	var (
-		stopCh chan struct{}
-		prev   common.Hash
-	)
+// interrupt aborts the in-flight sealing task.
+func (w *worker) interruptSealingTask() {
+	if w.prevTaskStopCh != nil {
+		close(w.prevTaskStopCh)
+		w.prevTaskStopCh = nil
+	}
+}
 
-	// interrupt aborts the in-flight sealing task.
-	interrupt := func() {
-		if stopCh != nil {
-			close(stopCh)
-			stopCh = nil
-		}
+func (w *worker) submitTaskToEngine(task *task) {
+	if w.newTaskHook != nil {
+		w.newTaskHook(task)
 	}
-	for {
-		select {
-		case task := <-w.taskCh:
-			if w.newTaskHook != nil {
-				w.newTaskHook(task)
-			}
-			// Reject duplicate sealing work due to resubmitting.
-			sealHash := w.engine.SealHash(task.block.Header())
-			if sealHash == prev {
-				continue
-			}
-			// Interrupt previous sealing operation
-			interrupt()
-			stopCh, prev = make(chan struct{}), sealHash
 
-			if w.skipSealHook != nil && w.skipSealHook(task) {
-				continue
-			}
-			w.pendingMu.Lock()
-			w.pendingTasks[w.engine.SealHash(task.block.Header())] = task
-			w.pendingMu.Unlock()
+	// Reject duplicate sealing work due to resubmitting.
+	sealHash := w.engine.SealHash(task.block.Header())
+	if sealHash == w.prevSealHash {
+		return
+	}
+	// Interrupt previous sealing operation
+	w.interruptSealingTask()
+	w.prevTaskStopCh, w.prevSealHash = make(chan struct{}), sealHash
 
-			if err := w.engine.Seal(w.chain, task.block, w.resultCh, stopCh); err != nil {
-				log.Warn("Block sealing failed", "err", err)
-			}
-		case <-w.exitCh:
-			interrupt()
-			return
-		}
+	if w.skipSealHook != nil && w.skipSealHook(task) {
+		return
+	}
+
+	w.pendingMu.Lock()
+	w.pendingTasks[w.engine.SealHash(task.block.Header())] = task
+	w.pendingMu.Unlock()
+
+	if err := w.engine.Seal(w.chain, task.block, w.resultCh, w.prevTaskStopCh); err != nil {
```

Comment:
i imagine this can be refactor to something like:

```go
ctx := // create context (save in to local field so we can cancel it )
go func() {
  sealedBlock, err := engine.Seal(ctx)
  // check ctx
  // do whatever we do on resultLoop()
}()


```